### PR TITLE
Measurements thread

### PR DESCRIPTION
**Pull request by DevPatel17**
- Functions for value conversion added
- lut for temperature conversion added
- moved isr function to global scope
- updated ioc file with latest pinout

### CODE REVIEW

File: STM32F405RGTx/Core/Src/rtos_threads/measurements_thread.cpp
Comment (DevPatel17) [inline STM32F405RGTx/Core/Src/rtos_threads/measurements_thread.cpp]: Realistically, the voltage here will never be 3.3V (or higher) but if there is we will have a division by zero exception. Is it worth adding an assertion before dividing?

File: STM32F405RGTx/Core/Inc/lut.hpp
Comment (RyanMah123) [inline STM32F405RGTx/Core/Inc/lut.hpp:6]: This is declared here, but is never defined anywhere.

File: STM32F405RGTx/Core/Src/rtos_threads/measurements_thread.cpp
Comment (RyanMah123) [reply STM32F405RGTx/Core/Src/rtos_threads/measurements_thread.cpp]: This conversion should also be done by the LUT. The LUT should do a direct conversion from an ADC value to a temperature.

File: STM32F405RGTx/Core/Src/rtos_threads/measurements_thread.cpp
Comment (RyanMah123) [inline STM32F405RGTx/Core/Src/rtos_threads/measurements_thread.cpp]: Again, these conversions should be done by the LUT.

If you think about what's happening in this code, `ADC_TO_VOLTAGE` will return a float, since its a voltage (although you are implicitly truncating the decimals since `val` is declared as an integer). Then you are trying to index an array with a float. Does this make sense?

The conversions from ADC to voltage, and voltage to resistance, then finally resistance to temperature ALL needs to be done in the lookup table, if we're going to be using a lookup table for this (which we should be imo).

File: scripts/generate_lut.py
Comment (RyanMah123) [inline scripts/generate_lut.py]: Is this the equation that you're going to use to generate the LUT? Like I've said, you can just straight up put in the beta equation here.

To give you an idea of what this `LUT_FUNC` would look like:

```Python
np.exp(...whatever...)*adc_to_voltage_coeff*voltage_to_temp_coeff
```

This will make it so your LUT converts directly from ADC to temperature.

File: STM32F405RGTx/Core/Src/rtos_threads/measurements_thread.cpp
Comment (RyanMah123) [inline STM32F405RGTx/Core/Src/rtos_threads/measurements_thread.cpp]: Whenever you define a macro like this, the argument for the macro should be enclosed by brackets, since all that happens with a macro is glorified copy and paste:

```C
#define ADC_TO_VOLTAGE(adc)    ( (adc)*(3.3/(1 << 12)) )
```

File: STM32F405RGTx/Core/Src/rtos_threads/measurements_thread.cpp
Comment (RyanMah123) [inline STM32F405RGTx/Core/Src/rtos_threads/measurements_thread.cpp]: Where does `17.25` come from? If it makes the code more readable, putting the calculation in the macro won't make it less efficient, since the calculation will be optimized out anyways.

File: STM32F405RGTx/Core/Inc/mc.hpp
Comment (RyanMah123) [inline STM32F405RGTx/Core/Inc/mc.hpp:29]: I know this isn't in the class diagram, but maybe it's a good idea to have two different variables for speed: 1 for speed measured by the IMU, and another for the speed measured by the ring encoder?

**Review by RyanMah123 — CHANGES_REQUESTED** [review]

File: STM32F405RGTx/Core/Inc/lut.hpp
Comment (DevPatel17) [reply STM32F405RGTx/Core/Inc/lut.hpp:6]: The correct lut is now added

File: STM32F405RGTx/Core/Inc/mc.hpp
Comment (DevPatel17) [reply STM32F405RGTx/Core/Inc/mc.hpp:29]: I was thinking that I'll add that once we also have the accelerometer code ready

File: STM32F405RGTx/Core/Src/rtos_threads/measurements_thread.cpp
Comment (RyanMah123) [inline STM32F405RGTx/Core/Src/rtos_threads/measurements_thread.cpp]: You also need to divide by the drv8323 differential amplifier gain

File: STM32_F405_MC/Core/Src/lut/ADC_TO_TEMP_LUT.c
Comment (RyanMah123) [inline STM32_F405_MC/Core/Src/lut/ADC_TO_TEMP_LUT.c]: Move this file to the correct location pls. Rn it's it `STM32_F405_MC/Core/Src/lut/ADC_TO_TEMP_LUT.c`.

It should be somewhere in `STM32F405RGTx/Core/Src`

File: STM32F405RGTx/Core/Src/rtos_threads/measurements_thread.cpp
Comment (RyanMah123) [inline STM32F405RGTx/Core/Src/rtos_threads/measurements_thread.cpp]: These should be handled in a separate function than the ADC for efficiency reasons.

File: STM32F405RGTx/Core/Src/rtos_threads/measurements_thread.cpp
Comment (RyanMah123) [inline STM32F405RGTx/Core/Src/rtos_threads/measurements_thread.cpp]: It won't let me write a comment on the actual function, but `MeasurementsThread::runMeasurements` needs to be implemented before this PR can be merged in. It needs to do the following in the while loop:

1. Process acceleration data
2. If the os flag 0x01 is set, process the data

**Review by RyanMah123 — DISMISSED** [review]

**Comment by RyanMah123** [comment]
Also, you should actually create the thread in `mc.cpp`

File: STM32F405RGTx/Core/Src/rtos_threads/measurements_thread.cpp
Comment (DevPatel17) [reply STM32F405RGTx/Core/Src/rtos_threads/measurements_thread.cpp:86]: Does this mean this PR can't be merged until we merge the accelerometer code? (including the implemented filter)

File: STM32F405RGTx/Core/Src/rtos_threads/measurements_thread.cpp
Comment (GuyBlumenthal) [inline STM32F405RGTx/Core/Src/rtos_threads/measurements_thread.cpp:114]: Just an implementatoin question, is the goal to see if the flag is set and if it is, process data, otherwise just go back to acceleration or are you supposed to wait for the flag to be set and then process data before going back to acceleration?

In the second one, the acceleration data measurements will be blocked until the flag is set essentially, so just wondering if this is fine or not.

File: STM32F405RGTx/Core/Src/rtos_threads/measurements_thread.cpp
Comment (GuyBlumenthal) [inline STM32F405RGTx/Core/Src/rtos_threads/measurements_thread.cpp:109]: Refer to line 90 - 92, should we be calling the error handler here if the status is not HAL_OK? That's the standard for everything i've seen, unless its ok if we don't initialize this

File: STM32F405RGTx/Core/Src/rtos_threads/measurements_thread.cpp
Comment (GuyBlumenthal) [inline STM32F405RGTx/Core/Src/rtos_threads/measurements_thread.cpp]: You need break statements in the switch case

File: STM32F405RGTx/Core/Src/rtos_threads/measurements_thread.cpp
Comment (DevPatel17) [reply STM32F405RGTx/Core/Src/rtos_threads/measurements_thread.cpp:114]: I think it makes most sense to go with the first. There's no point in waiting for the flag to be set before getting acceleration data.

File: STM32F405RGTx/Core/Src/rtos_threads/measurements_thread.cpp
Comment (GuyBlumenthal) [inline STM32F405RGTx/Core/Src/rtos_threads/measurements_thread.cpp]: You can remove this printf. Generally, we're not going to be connected over UART so there's no reason to have it in. If we're debugging over UART, you can just add prints as you need.

File: STM32F405RGTx/Core/Src/rtos_threads/measurements_thread.cpp
Comment (GuyBlumenthal) [inline STM32F405RGTx/Core/Src/rtos_threads/measurements_thread.cpp:31]: How did you decide on the stack size here?

File: STM32F405RGTx/STM32F405RGTx.ioc
Comment (GuyBlumenthal) [inline STM32F405RGTx/STM32F405RGTx.ioc:132]: Are you trying to set the DMA interrupt on a higher priority here or lower priority? A lower number results in a higher priority so in this case you're decreasing priority, i.e. everything else will run before the dma interrupt

Comment (DevPatel17) [reply STM32F405RGTx/STM32F405RGTx.ioc:132]: I am trying to set it to a higher priority here
The DMA priority was set to "very high" through  cubeide and I noticed that bms also has it set to this as well despite the .ioc in cube ide showing "high priority"

Should I change the 5 to a 0 instead?

File: STM32F405RGTx/Core/Src/rtos_threads/measurements_thread.cpp
Comment (DevPatel17) [reply STM32F405RGTx/Core/Src/rtos_threads/measurements_thread.cpp:31]: It was taken off of the bms measurements thread since we should have a similar stack

Comment (GuyBlumenthal) [reply STM32F405RGTx/Core/Src/rtos_threads/measurements_thread.cpp:31]: Okie

File: STM32F405RGTx/STM32F405RGTx.ioc
Comment (GuyBlumenthal) [reply STM32F405RGTx/STM32F405RGTx.ioc:132]: Hmm, idk maybe this is a question for Ryan. For now you can leave it tho since I doubt we'll have timing issues

**Review by GuyBlumenthal — APPROVED** [review]
Ok, it looks good to me. There's like a few things that aren't super crucial that just need to be followed up with after. Also, **make sure to pull from develop before merging**. You're two commits behind from develop

**Pull request merged**
